cdap-gateway/src/main/java/io/cdap/cdap/gateway/router/handlers/AuthenticationHandler.java

```diff
@@ -108,43 +107,57 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
       ctx.fireChannelRead(msg);
       return;
     }
-    TokenState tokenState = validateAccessToken(request, ctx.channel());
-    if (tokenState.isValid()) {
-      ctx.fireChannelRead(msg);
-      return;
-    }
-
-    // Response with failure, plus optionally audit log
+    UserIdentityPair userIdentityPair;
     try {
-      HttpHeaders headers = new DefaultHttpHeaders();
-      JsonObject jsonObject = new JsonObject();
-      if (tokenState == TokenState.MISSING) {
-        headers.add(HttpHeaderNames.WWW_AUTHENTICATE, String.format("Bearer realm=\"%s\"", realm));
-        LOG.debug("Authentication failed due to missing token");
-      } else {
-        headers.add(HttpHeaderNames.WWW_AUTHENTICATE,
-                    String.format("Bearer realm=\"%s\" error=\"invalid_token\"" +
-                                    " error_description=\"%s\"", realm, tokenState.getMsg()));
-        jsonObject.addProperty("error", "invalid_token");
-        jsonObject.addProperty("error_description", tokenState.getMsg());
-        LOG.debug("Authentication failed due to invalid token, reason={};", tokenState);
+      userIdentityPair = userIdentityExtractor.extract(request);
+      // User identity extraction succeeded, so set some header properties and allow the call through
+      request.headers().remove(HttpHeaderNames.AUTHORIZATION);
+      String credential = userIdentityPair.getUserCredential();
+      // For backwards compatibility, we continue propagating credentials by default. This may change in the future.
+      if (cConf.getBoolean(Constants.Security.Authentication.AUTH_PROPAGATE_USER_CREDENTIAL, true) &&
```

Comment:
In our case, should backwards compatibility be a concern? If somebody has a cdap-site.xml file which does not have this key but has auth enabled, the behavior would be different if they upgraded versions.

Reply:
I went ahead and set a default in cdap-default.xml to be false because this should be turned off by default in future versions anyway.

Reply:
What is the correct default behavior if someone upgraded without any changes in the cdap-site.xml?

Reply:
The default (previous) behavior is that tokens are propagated, so this needs to be true to maintain the default behavior. I added this config because I'm not entirely convinced that the default behavior is correct, but we can completely remove the config if necessary.

Reply:
Let's have the config to preserve the old behavior.

Reply:
Understood, sounds good